### PR TITLE
added forEachObjIndexed

### PR DESCRIPTION
**Pull request by beenotung**
ref: http://ramdajs.com/docs/#forEachObjIndexed

### CODE REVIEW

**Comment by beenotung** [comment]
Build failed:
```
linting...
index.d.ts[909, 1]: trailing whitespace
```
No ideas about it, cannot see the trailing whitespace manually.

The build was failed before I edit the file, I guess I am not making it even worse.

File: index.d.ts [hunk 1]
```diff
@@ -906,6 +906,12 @@ declare namespace R {
         forEach<T>(fn: (x: T) => void, list: List<T>): T[];
         forEach<T>(fn: (x: T) => void): (list: List<T>) => T[];
         // forEach<T>: CurriedFunction2<(x: T) => void, List<T>, T[]>;
+  
+        /**
+         * Iterate over an input object, calling a provided function fn for each key and value in the object.
+         */
+        forEachObjIndexed<A>(fn: (val: A, key: string|number) => void, o: Obj<A>): Obj<A>;
+        forEachObjIndexed<A>(fn: (val: A, key: string|number) => void): (o: Obj<A>) => Obj<A>;
```
Comment (tycho01) [inline index.d.ts]: If we're adding `number` to support the case of mapping arrays to objects, use `Prop` over `string|number`?

**Comment by tycho01** [comment]
Yeah no worries, there are still a bunch of failing tests including issues caused by big issues in TypeScript itself; I'd be ready to merge, just had a minor nitpick.
To confirm though, had you added number to map over objects with numerical keys (would turn into strings) or over arrays (which is entirely possible though maybe uncommon usage with the result would become an object)?

**Comment by BTBurke** [comment]
I believe this should only take string-indexed maps, a.k.a `Dictionary<T>` and should also add the optional third parameter which is the complete object per the Ramda docs which give the function signature as `((a, String, StrMap a) → Any) → StrMap a → StrMap a`:

```
forEachObjIndexed<T>(fn: (val: T, key: string, obj?: Dictionary<T>) => void, o: Dictionary<T>): Dictionary<T>;
forEachObjIndexed<T>(fn: (val: T, key: string, obj?: Dictionary<T>) => void): (o: Dictionary<T>) => Dictionary<T>;
```

I patched mine manually with this sig and it's working fine.

**Comment by beenotung** [comment]
@BTBurke nice, shall I close this request?

**Comment by tycho01** [comment]
@BTBurke: feel free to submit PRs as well! :)

So no matter if the input is an object with 'string-like' keys (strings), 'number-like' keys (strings), or an array (numbers), it seems the key parameter always turns out as a string. So that case seems clear-cut.

On `A` vs. `T`, I'd recently noticed the Sanctuary typings had been neatly following their docs in terms of generics names, which is kinda pretty, although this repo (starting with those before me, enriched and donnut) had traditionally been following TS-ish conventions of using `T` and the like. I do see the appeal of staying close to the Ramda docs, but if we're gonna do it then maybe we could do it all at once at one point to kinda keep things internally consistent.

I think the `forEachObjIndexed<T>(fn: (val: T, key: string, obj?: Dictionary<T>) => void, o: Dictionary<T>): Dictionary<T>;` version could still be improved in one respect though -- considering the case of users supplying an array rather than an object, which actually works [fine](http://ramdajs.com/repl/?v=0.23.0#?var%20printKeyConcatValue%20%3D%20%28value%2C%20key%29%20%3D%3E%20console.log%28JSON.stringify%28key%29%20%2B%20%27%3A%27%20%2B%20JSON.stringify%28value%29%29%3B%0AR.forEachObjIndexed%28printKeyConcatValue%2C%20%5B1%2C%202%5D%29%3B%20%2F%2F%3D%3E%20%7Bx%3A%201%2C%20y%3A%202%7D%0A%2F%2F%20logs%20x%3A1%0A%2F%2F%20logs%20y%3A2):
```
var printKeyConcatValue = (value, key) => console.log(JSON.stringify(key) + ':' + JSON.stringify(value));
R.forEachObjIndexed(printKeyConcatValue, [1, 2]);
// logs: "0":1
// logs: "1":2
// returns: [1, 2]
```

At this point you may note that I'm going by its implementation, rather than by the way its described in the docs.
I'm definitely open to opinions as to what would be most desirable there, but so far I'll admit I've been doing this for other functions as well, actually testing what's possible, then generalizing over that to allow the typings to accommodate any legitimate usage of the library, whether described in the docs or not.

(Context: this had been notably relevant in generalizing `Array` cases to `ArrayLike`, which also allows string inputs. This actually turned out to be the case for a vast amount of Ramda functions, and allowed keeping one type definition rather than separately specifying array vs. string in those cases.)

I'm under the impression it'd be unfortunate to discourage array inputs here, as that particular use-case (array, side-effects, key parameters, curried) is not well-covered by other functions.

tl;dr I'm happily merging any suggestions here, though I'd like to make it broader than `Dictionary`. That said I'm terrible with generics names so I'm totally cool with PRs to improve on even those!

**Comment by BTBurke** [comment]
I'll work up a PR for this.  I'm in a little bit of disagreement about
having the types conform to the implementation (in JS) rather than
trying to enforce the types that are listed in the Ramda
documentation.  When I look at the implementation, it seems you get the
behavior you mentioned because some browsers implement
`Array.prototype.keys()` to return the index as a string for an array
input.  That kind of behavior is one of the reasons that I use
Typescript in the first place because I don't want JS to silently
coerce types in a non-intuitive way.  That behavior may also be browser
dependent, but I'm not certain about that.
Ramda already has a `forEach` method that applies to arrays and the
`addIndex` operation to generate a function that would give you a
similar type of map with the index and the value passed to your
callback function for array inputs.  Personally, I'd like to see the
type definitions stay as close to the docs as possible so I get
compilation errors when I'm incorrectly applying the function.  I think
the principle of least surprise is to keep the type definitions as
close to the docs as possible so we don't get hard-to-debug errors that
result from the type definitions diverging from the Ramda docs.
I'll work on a PR after looking more at whether there are any
potentially bad side effects from replacing `Dictionary<T>` with
something that would allow generic array-like structures.  It would
likely be easiest to implement with a union type of `Dictionary<T> |
Array<T>`.

**Comment by tycho01** [comment]
If I might add some reasoning to my position there:
- the actual working of Ramda is much more consistent than their documentation (notably for the cases where both array and string are supported)
- generalizing over the actual working actually simplifies typings compared to the alternative of considering them separate where applicable
- wanting to use this function on arrays is kind of legitimate in that there is no alternative offered on par with it in terms of convenience (`forEach` with `addIndex`)

> I'll work on a PR after looking more at whether there are any potentially bad side effects from replacing `Dictionary<T>` with something that would allow generic array-like structures.  It would likely be easiest to implement with a union type of `Dictionary<T> | Array<T>`.

Yeah, this is what I'd done now with `Struct<T>` .

I wouldn't mind input from others, but I'm growing more sympathetic to your arguments. This raises similar questions on whether it's useful to allow support for strings wherever Ramda (implictly) allows them.
Honestly the lines can get blurry for some cases on whether the Ramda docs should better document its behavior or we should change things here, but yeah. I'm definitely open to this discussion.
